### ISSUE TITLE
"SWS/S&M: Map selected tracks MIDI input to channel x" doesn´t work during recording

### ISSUE THREAD

**Issue by jurek123** [reaper-oss/sws#913]

**Comment by veto-gh** [comment]
Thanks for the report, I can reproduce it.

@Devs:
This line is preventing mapping channels during recording:
https://github.com/reaper-oss/sws/blob/400c3c13949805afc37b87627a0c8664f436508f/SnM/SnM_ChunkParserPatcher.h#L260

So it has to be with all actions which involve chunk editing during recording.
Alas I have no clue why Jeffos added this check, but I suppose he had a good reason (for security reasons to not possibly crash the project during recording?).

Does anybody know more?

**Comment by cfillion** [comment]
The GetPlayState test was added in 9f2726a056137a679ee91a62e74f2780b9b33747:

> Global "Safety" update (temporary, I hope): many S&M actions will have no effect while *recording*.
See http://forum.cockos.com/showthread.php?t=52002 for tech. details

Prior to c1a4f8f987cfd743ad4d31546d6269671c9659d9 there was a comment saying:

> // Commit: only if needed, i.e. if m_object has been set and if the chunk has been updated
> // note: global protections here (temporary, I hope). In REAPER v3.66, it's safer NOT to 
> // patch while recording and to remove all ids before patching, see http://forum.cockos.com/showthread.php?t=52002

**Comment by veto-gh** [comment]
Nice, thanks for the additional info @cfillion !

The [last Post](https://forum.cockos.com/showpost.php?p=559957&postcount=22) in this 2010 thread makes me tend to remove the check but I'm sure Jeffos still had very good reasons to leave it in up until last year.
I would leave it in too, but maybe we could rename relevant actions adding "... (no-op during recording)". Not sure how many that would be though..

**Comment by nofishonfriday** [comment]
Asked Justin about it, let's see what he says...
https://www.askjf.com/index.php?q=3889s

**Comment by veto-gh** [comment]
Thank you.
I realized the Call Hierarchy of ParsePatch(), which I guess is the main method for chunk patching, is _huge_. Jeffos (and others) made good use of his swiss army knife tool.
These are a couple of things I found after a quick search that are effected by this check:
- Midi CC lane actions and Reascript functions
- Send/Receive/Cue Buss actions and Reascript functions
- many Item & Track FX actions
- Reassign MIDI learned channels of ...
- Show/Hide take vol/pan/mute/pitch envelope
- actions involving Track Templates
- functions used by SNM Resources and Live Configs
- ...

**Comment by nofishonfriday** [comment]
Thanks for checking veto.

Seeing it affects so much I'd say safety first and leave the check in (so the issue would be a "Live with it.")
But that's just me of course...

**Comment by veto-gh** [comment]
I'm pretty much a noob but I still tend to leave it in too (at least for the older actions/functions).
Maybe a long or separate beta would make sense too.

Concerning the OP, if he doesn't like the JSFX approach, I'm ready to do it by Reascript for him.

**Comment by jurek123** [comment]
Indeed, I´d be very interested if you could do a Reascript for that? That´d be fantastic.

Jurek (OP)

**Comment by veto-gh** [comment]
Hi Jurek,
for different channels duplicate the script, change the file name and edit the source (look for the "user settings" part).
If you set CHANNEL to "nil" (without quotes) you can map MIDI to the source channel.

[Map selected tracks MIDI input to channel 1.lua.zip](https://github.com/reaper-oss/sws/files/1438936/Map.selected.tracks.MIDI.input.to.channel.1.lua.zip)

**Comment by veto-gh** [comment]
Hi Jurek,
for different channels duplicate the script, change the file name and edit the source (look for the "user settings" part).
If you set CHANNEL to 0 you can map MIDI to the source channel.

File: [Map selected tracks MIDI input to channel 1.lua.zip](https://github.com/reaper-oss/sws/files/1439283/Map.selected.tracks.MIDI.input.to.channel.1.lua.zip)

**Comment by jurek123** [comment]
Hey, I just had a chance to test it, works like a charm! Thank you so much!

Jurek

**Comment by nofishonfriday** [comment]
Since OP's original issue sounds solved I took the liberty to rename the title to reflect the broader topic going on here.
